src/main/MainTabsScreen.js

```diff
@@ -117,32 +117,46 @@ function MainTabsScreen(props: Props) {
   );
 }
 
-// `connect` does something useful for us that `useSelector` doesn't
-// do: it interposes a new `ReactReduxContext.Provider` component,
-// which proxies subscriptions so that the descendant components only
-// rerender if this one continues to say their subtree should be kept
-// around. See
-//   https://github.com/zulip/zulip-mobile/pull/4454#discussion_r578140524
-// and some discussion around
-//   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/converting.20to.20Hooks/near/1111970
-// where we describe some limits of our understanding.
-//
-// We found these things out while investigating an annoying crash: we
-// found that `mapStateToProps` on a descendant of `MainTabsScreen`
-// was running -- and throwing an uncaught error -- on logout, and
-// `MainTabsScreen`'s early return on `!haveServerData` wasn't
-// preventing that from happening.
-export default connect(state => ({ haveServerData: getHaveServerData(state) }))((props: Props) =>
-  props.haveServerData ? (
-    <MainTabsScreen {...props} />
-  ) : (
-    // Show a full-screen loading indicator while waiting for the
-    // initial fetch to complete, if we don't have potentially stale
-    // data to show instead. Also show it for the duration of the nav
-    // transition just after the user logs out (see our #4275).
-    //
-    // And avoid rendering any of our main UI, to maintain the
-    // guarantee that it can all rely on server data existing.
-    <FullScreenLoading />
-  ),
-);
+/**
+ * A HOC for any server-data-dependent screen component that might be
+ *   mounted when we don't have server data.
+ *
+ * Prevents rerendering of the component's subtree unless we have
+ * server data.
+ *
+ * Passing a `dispatch` prop or a `haveServerData` prop to the
+ * returned component will lead to undefined behavior; don't.
```

Comment:
Will it? I believe what will happen is `connect` will just clobber those props.

Also I think if Flow sees you're doing that, you'll get an error because the returned component-type doesn't accept those props -- it accepts what the argument `Comp` accepts, minus those two. That Flow error will probably be pretty obscure, though.

Reply:
> Will it? I believe what will happen is `connect` will just clobber those props.

Ah, I meant "undefined" in the sense of "we don't care to define", not "we can't define". 🙂 To make the component's interface more stable across implementation changes.

Does that sound like a good strategy? If so, I think this bit of jsdoc might be good to keep, especially with the likelihood that the Flow error would be obscure. There may be a better way to write it, though.

Reply:
Hmm, rereading, I think what I said above is correct about the runtime behavior -- but if the inner component wants to take those props, and you pass them to the outer component, Flow won't give an error.

That's seen in the fact that the signature takes a component and returns a component which takes the same props:
```js
export default function withHaveServerDataGate<P: { ... }, C: ComponentType<$Exact<P>>>(
  Comp: C,
): ComponentType<$Exact<ElementConfig<C>>> {
```
so if Flow were going to complain about how the implementation doesn't totally agree with that, it should do so already when type-checking the implementation.

And indeed empirically:
```js
function example(c: ComponentType<{| dispatch: Dispatch |}>) {
  const MyComponent = withHaveServerDataGate(c);
  <MyComponent />; // error
  <MyComponent dispatch={(null: $FlowFixMe)} />; // no error
}
```

I think that's basically OK, because trying to pass these two props through doesn't seem like a likely mistake to make, and I don't think one would get very far before seeing it wasn't working.

But yeah, a jsdoc mention of it seems good. I'll try tweaking the wording.